### PR TITLE
Update NumPy installation on Mac OS X Travis build

### LINKED ISSUE #1822: MacOS stage of Travis CI is broken
This looks like an issue with a Python 3.8 vs. 3.9 version conflict. Possibly, this is due to the default Python version in the Travis CI MacOS image having been upgraded to 3.9.

### PR DESCRIPTION
**Pull request by lekshmideepu**
Fixing the issues with cython,numpy and matplotlib on OsX under Travis after the upgrade of Python to 3.9.0
Fix #1822

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -153,25 +151,27 @@ before_install:
...
          # terminaltables is required by parse_travis_log.py to create
          # the build summary.
-         pip install cython scipy matplotlib terminaltables junitparser mpi4py
-         pip install numpy==1.16.5
+         pip install scipy junitparser mpi4py
+         if [ "$TRAVIS_OS_NAME" != "osx" ]; then
+           pip install numpy nose
+           pip install cython matplotlib terminaltables
+         else
+           # Due to the issue with numpy C-extensions, have to stick to a speciic version on OsX
```
Comment (heplesser) [inline .travis.yml:169]: Could you add some more information, a link to an issue or a website about this problem? Just so we remember next year what this was about.

File: .travis.yml [hunk 2]
```diff
@@ -153,25 +151,27 @@ before_install:
...
-         pip install numpy==1.16.5
+         pip install scipy junitparser mpi4py
+         if [ "$TRAVIS_OS_NAME" != "osx" ]; then
+           pip install numpy nose
+           pip install cython matplotlib terminaltables
+         else
+           # Due to the issue with numpy C-extensions, have to stick to a speciic version on OsX	 
+           pip3 install numpy==1.19.3
+           pip3 install cython matplotlib terminaltables
+           wget https://arken.nmbu.no/~plesser/nose-1.3.7noman-py3-none-any.whl
```
Comment (heplesser) [inline .travis.yml:172]: It is not really ideal to depend on my personal university website here. Should we include this file in the NEST repo to have it available. I think it would also be helpful to keep the comment on why we don not just do a `pip install nose`.

```
# updating pip breaks on osx, must install man-page free nose
```

**Review by heplesser — CHANGES_REQUESTED** [review]
@lekshmideepu Thanks! It looks mostly good, but see my comments. I am afraid the tests will presently no make it to the OSX stage because of the strict test on the number of warnings.

File: .travis.yml [hunk 3]
```diff
@@ -1,6 +1,5 @@
 
 # .travis.yaml
-#
```
Comment (jougs) [inline .travis.yml:3]: Can you please revert this change to keep the file consistent with others?

File: .travis.yml [hunk 4]
```diff
@@ -74,7 +73,6 @@ addons:
      - *BASE_SOURCES
      - r-packages-bionic
```
Comment (jougs) [inline .travis.yml:74]: I think this is also not needed
```suggestion
```

File: .travis.yml [hunk 5]
```diff
@@ -153,25 +151,27 @@ before_install:
...
          # Installing additional packages using pip as they only have
          # outdated versions in the Travis package whitelist.
          # terminaltables is required by parse_travis_log.py to create
          # the build summary.
-         pip install cython scipy matplotlib terminaltables junitparser mpi4py
-         pip install numpy==1.16.5
+         pip install scipy junitparser mpi4py
+         if [ "$TRAVIS_OS_NAME" != "osx" ]; then
+           pip install numpy nose
+           pip install cython matplotlib terminaltables
```
Comment (jougs) [inline .travis.yml:167]: Not sure, but maybe we should use `pip3` everywhere instead of `pip`?

```suggestion
           pip3 install numpy nose
           pip3 install cython matplotlib terminaltables
```

**Review by jougs — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 1 above)
Comment (lekshmideepu) [reply .travis.yml:168]: Thanks for the review. Only after some trial and error I figured out that one specific version of numpy works on OsX. So I don't have any link for this issue. Instead I have https://numpy.org/devdocs/user/troubleshooting-importerror.html

File: .travis.yml (hunk 3 above)
Comment (lekshmideepu) [reply .travis.yml:3]: Thanks for the review. I would do that.

**Comment by heplesser** [comment]
@lekshmideepu I manually started the macOS test (it does not run automatically because the full-build test fails because of warnings-count), see https://travis-ci.org/github/nest/nest-simulator/jobs/743499027. All builds, but the nosetest install fails. I will make a suggestion as inline comment.

File: .travis.yml [hunk 6]
```diff
@@ -153,25 +150,28 @@ before_install:
            brew unlink gcc
            brew install gcc@8
            brew link gcc@8
+           brew uninstall numpy
```
Comment (heplesser) [inline .travis.yml:153]: I'd suggest to remove this again. The log shows that brew refuses to uninstall numpy (1.19.4 at present), because it is a dependency.

File: .travis.yml [hunk 7]
```diff
@@ -153,25 +150,28 @@ before_install:
...
+         if [ "$TRAVIS_OS_NAME" != "osx" ]; then
+           pip3 install numpy nose
+           pip3 install cython matplotlib terminaltables
+         else
+           # Due to the issue with numpy C-extensions, have to stick to a specific version on OsX
+           #https://numpy.org/devdocs/user/troubleshooting-importerror.html
+           pip3 install numpy==1.19.3
+           pip3 install cython matplotlib terminaltables nose
+           #wget https://arken.nmbu.no/~plesser/nose-1.3.7noman-py3-none-any.whl
+           #pip3 install nose-1.3.7noman-py3-none-any.whl
```
Comment (heplesser) [inline .travis.yml:173]: Nose installation fails because of some clash when trying to install man pages. Maybe installing nose user-only will help and avoid then need to use the "noman" archive:
```suggestion
           pip3 install cython matplotlib terminaltables
           
           # nose must be installed user-only to avoid permissions conflict for man-pages
           pip3 install --user nose
```

File: extras/travis_build.sh [hunk 8]
```diff
@@ -35,6 +35,8 @@ set -e
 if [[ $OSTYPE = darwin* ]]; then
     export CC=$(ls /usr/local/bin/gcc-* | grep '^/usr/local/bin/gcc-\d$')
     export CXX=$(ls /usr/local/bin/g++-* | grep '^/usr/local/bin/g++-\d$')
+    #Ensure that nosetests path can be found
+    export PATH=/usr/local/bin/nosetests:$PATH
```
Comment (heplesser) [inline extras/travis_build.sh:39]: Adjust to user-local installation:
```suggestion
    export PATH=$HOME/.local/bin:$PATH
```

**Review by heplesser — CHANGES_REQUESTED** [review]
@lekshmideepu I added some suggestions for user-local install of nosetests to avoid permissions problems when pip tries to install man pages.

File: extras/travis_build.sh [hunk 9]
```diff
@@ -35,6 +35,8 @@ set -e
 if [[ $OSTYPE = darwin* ]]; then
     export CC=$(ls /usr/local/bin/gcc-* | grep '^/usr/local/bin/gcc-\d$')
     export CXX=$(ls /usr/local/bin/g++-* | grep '^/usr/local/bin/g++-\d$')
+    #Ensure that nosetests path can be found
+    export PATH=$HOME/.local/bin:$PATH
```
Comment (heplesser) [inline extras/travis_build.sh:39]: The ways of installation are curious. The log of the last build shows a successful nose install, but to [yet another destination](https://travis-ci.org/github/nest/nest-simulator/jobs/743943712#L357), so this line should most likely be
```suggestion
    export PATH=/Users/travis/Library/Python/3.9/bin:$PATH
```
```

File: .travis.yml [hunk 10]
```diff
@@ -156,22 +153,26 @@ before_install:
...
+           pip3 install cython matplotlib terminaltables
+         else
+           # Due to the issue with numpy C-extensions, have to stick to a specific version on OsX
+           #https://numpy.org/devdocs/user/troubleshooting-importerror.html
+           pip3 install numpy==1.19.3
+           pip3 install cython matplotlib terminaltables
+           # nose must be installed user-only to avoid permissions conflict for man-pages
+           pip3 install --user  nose
+           #wget https://arken.nmbu.no/~plesser/nose-1.3.7noman-py3-none-any.whl
+           #pip3 install nose-1.3.7noman-py3-none-any.whl
```
Comment (heplesser) [inline .travis.yml:174]: Remove these two lines
```suggestion
```

**Review by heplesser — CHANGES_REQUESTED** [review]
@lekshmideepu Progress ... see some suggestions below, hopefully taking us yet further towards success ;).

**Comment by heplesser** [comment]
See also #1852, which will solve the problems with the currently failing "full" build.

**Comment by jougs** [comment]
@lekshmideepu: Please pull master. The problems with the warnings should be gone now that #1852 is merged.

File: .travis.yml (hunk 6 above)
Comment (lekshmideepu) [reply .travis.yml]: Sure!

**Review by heplesser — APPROVED** [review]

File: .travis.yml [hunk 11]
```diff
@@ -1,6 +1,5 @@
 
 # .travis.yaml
```
Comment (jougs) [inline .travis.yml:2]:
```suggestion
# .travis.yaml

```

**Review by jougs — APPROVED** [review]
I'm fine once Travis is. I'll commit my minor suggestion right before merging.

**Comment by heplesser** [comment]
All green again 👍 !!!

**Comment by lekshmideepu** [comment]
Finally :-) Thanks for the reviews @heplesser  and @jougs 👍
